src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Models/RazorSemanticTokensLegend.cs

```diff
@@ -34,44 +33,57 @@ internal class RazorSemanticTokensLegend
-    public static int RazorDirectiveAttribute => TokenTypesLegend[RazorDirectiveAttributeType];
-    public static int RazorDirectiveColon => TokenTypesLegend[RazorDirectiveColonType];
-    public static int RazorTagHelperAttribute => TokenTypesLegend[RazorTagHelperAttributeType];
-    public static int RazorTagHelperElement => TokenTypesLegend[RazorTagHelperElementType];
-    public static int RazorTransition => TokenTypesLegend[RazorTransitionType];
-
-    public static int CSharpKeyword => TokenTypesLegend["keyword"];
-    public static int CSharpOperator => TokenTypesLegend["operator"];
-    public static int CSharpPunctuation => TokenTypesLegend["punctuation"];
-    public static int CSharpString => TokenTypesLegend["string"];
-    public static int CSharpVariable => TokenTypesLegend["variable"];
-
-    // C# types + Razor types
-    public static readonly ImmutableArray<string> TokenTypes = GetTokenTypes();
-
-    private static ImmutableArray<string> GetTokenTypes()
+    public int MarkupAttribute => _razorTokenTypeMap[MarkupAttributeType];
+    public int MarkupAttributeQuote => _razorTokenTypeMap[MarkupAttributeQuoteType];
+    public int MarkupAttributeValue => _razorTokenTypeMap[MarkupAttributeValueType];
+    public int MarkupComment => _razorTokenTypeMap[MarkupCommentType];
+    public int MarkupCommentPunctuation => _razorTokenTypeMap[MarkupCommentPunctuationType];
+    public int MarkupElement => _razorTokenTypeMap[MarkupElementType];
+    public int MarkupOperator => _razorTokenTypeMap[MarkupOperatorType];
+    public int MarkupTagDelimiter => _razorTokenTypeMap[MarkupTagDelimiterType];
+    public int MarkupTextLiteral => _razorTokenTypeMap[MarkupTextLiteralType];
+
+    public int RazorComment => _razorTokenTypeMap[RazorCommentType];
+    public int RazorCommentStar => _razorTokenTypeMap[RazorCommentStarType];
+    public int RazorCommentTransition => _razorTokenTypeMap[RazorCommentTransitionType];
+    public int RazorComponentAttribute => _razorTokenTypeMap[RazorComponentAttributeType];
+    public int RazorComponentElement => _razorTokenTypeMap[RazorComponentElementType];
+    public int RazorDirective => _razorTokenTypeMap[RazorDirectiveType];
+    public int RazorDirectiveAttribute => _razorTokenTypeMap[RazorDirectiveAttributeType];
+    public int RazorDirectiveColon => _razorTokenTypeMap[RazorDirectiveColonType];
+    public int RazorTagHelperAttribute => _razorTokenTypeMap[RazorTagHelperAttributeType];
+    public int RazorTagHelperElement => _razorTokenTypeMap[RazorTagHelperElementType];
+    public int RazorTransition => _razorTokenTypeMap[RazorTransitionType];
+
+
+    public SemanticTokensLegend Legend => _legend;
+
+    private readonly SemanticTokensLegend _legend;
+    private readonly Dictionary<string, int> _razorTokenTypeMap;
+
+    public RazorSemanticTokensLegend(ClientCapabilities clientCapabilities)
     {
         var builder = ImmutableArray.CreateBuilder<string>();
```

Comment:
Since you're in here, would you mind switching this to use a pooled array builder?